LED state detection with OpenCV

4c3y opened the pull request:
- Created time_stamper_cv catkin package
- Implemented LED state detection with OpenCV
- Implemented visualization
- Configurable from ROS
- Unittests

4c3y committed:
Created catkin package time_stamper_cv, Added opencv hello world

4c3y committed:
Merge branch 'main' into feature/cv

4c3y committed:
Added passthrough ros node, and show image

4c3y committed:
Added BlobDetectionTest.cpp

4c3y committed:
Added image undistort calibration, Added working blob detector in node + launch files

4c3y committed:
Added first working corner led detection

4c3y committed:
Added gtest

4c3y committed:
Added convex hull with visualisation

4c3y committed:
Converted camera config to yaml

4c3y committed:
Added Trigonometry.h, Replaced test_algorithms.cc with test_trigonometry.cc, Added compiler flags

4c3y committed:
Added catkin_simple in package.xml

4c3y committed:
Fixed CMake for CI

4c3y committed:
Added inner angle calculation for each point in hull vector + visualisation

4c3y committed:
Removed unused code in Main.cpp

4c3y committed:
Filter invalid shapes

4c3y committed:
Added ShapeValidation

4c3y committed:
Removed old corner led detection, Fixed Node::Filter + Refactoring

4c3y committed:
Fixed ros log spam

4c3y committed:
Converted std::vector<double> to std::vector<PointAngle> in ShapeValidation, Updated unittests

4c3y committed:
Added warpPerspective with calculated homography, Created subimage of first led

4c3y committed:
Removed debug and simplified code

4c3y committed:
Calculated average brightness of first led in cropped image

4c3y committed:
Calculated average brightness of every led, added binary counter detection

4c3y committed:
Extracted methods convertKeyPoints(), createBlobDetectorPtr(), visualizeCorners()

4c3y committed:
Refactoring, Added counter and shape validity on visualization_mat

4c3y committed:
Renamed ConvexShape from ShapeValidation, Refactored node

4c3y committed:
Added ConvexShape object with extracted methods

4c3y committed:
Improved ConvexShape object performance

4c3y committed:
Refactored static methods in ConvexShape

4c3y committed:
Renamed rotateVector() to calculateRotatedVector() + reworked logic

4c3y committed:
Renamed convertPointAngles() to getRotatedPointAngles2f() and moved from Node to ConvexShape

4c3y committed:
Moved image processing logic from Node to Calibration, Added Configuration.h

4c3y committed:
Refactored ConvexShape

4c3y committed:
Bugfixed and Refactored Node

4c3y committed:
Refactored Calibration

4c3y committed:
Added Calibration::Visualize()

4c3y committed:
Added Calibration::GenerateLedRow(), Merged visualization screens + Refactoring

4c3y committed:
Added GetLedCounter(), improved GenerateLedRow()

4c3y committed:
Refactored Trigonometry.cpp

4c3y committed:
Disabled test_shape_validation.cc temporarily

michaelpantic committed:
Fixes for 18.04

4c3y committed:
Reduced ConvexShape memory reallocation and converted to member variable in Calibration.h

4c3y committed:
Moved Filter() from Node to ConvexShape

4c3y committed:
Added Common.h + Applied typedefs in Calibration

4c3y committed:
Added tolerance filter

4c3y committed:
Renamed test_shape_validation to test_convex_shape.cc, reworked unittests

4c3y committed:
Added time_stamper_cv_node in launch file + refactoring

4c3y committed:
Removed BlobDetectionTest

4c3y committed:
Fixed ConvexShape::calculateSortedPointAngles()

4c3y committed:
Extracted LedParser from Calibration, Fixed Common.h

4c3y committed:
Improved LedParser performance + refactoring

4c3y committed:
Bugfixed and removed debug in Calibration.cpp

4c3y committed:
Added test_led_parser.cc

4c3y committed:
Updated README.md

4c3y committed:
Improved ConvexShape::calculatePointAngles()

4c3y committed:
Added documentation in multiple headers

4c3y committed:
Added docs in LedParser.h + refactoring

4c3y committed:
Fixed typo

4c3y committed:
Added docs in Trigonometry.h, made Trigonometry::LawOfCos() private

4c3y committed:
Fixed Unittests

4c3y committed:
Make visualization configurable from ROS

michaelpantic commented on time_stamper_cv/src/Node.cpp:
```diff
@@ -0,0 +1,55 @@
+#include "Node.h"
+#include "Trigonometry.h"
+
+Node::Node() {
```
Please rename Node to a more verbose name

michaelpantic commented on time_stamper_cv/src/Node.cpp:
```diff
@@ -0,0 +1,55 @@
+bool Node::Init() {
+  img_pub_ = nh_.advertise<sensor_msgs::Image>("time_stamper_cv_image", 1);
+  calibration_->SetVisualization(nh_private_.param("show_visualization", true));
+  return true;
```
no need for return value

michaelpantic commented on time_stamper_cv/src/Node.cpp:
```diff
@@ -0,0 +1,55 @@
+}
+
+void Node::CallbackRawImage(const sensor_msgs::Image &image) {
+  cv_bridge::CvImage out_msg = calibration_->ProcessImage(image);
```
maybe rename calibration_

michaelpantic commented on time_stamper_cv/include/Calibration.h:
```diff
@@ -0,0 +1,56 @@
+
+class ConvexShape;
+
+//TODO refactoring
```
do it or remove todo

michaelpantic commented on time_stamper_cv/include/Calibration.h:
```diff
@@ -0,0 +1,56 @@
+
+  std::vector<cv::KeyPoint> keypoints_{};
+  cv::Ptr<cv::SimpleBlobDetector> detector_;
+  bool visualization_ = false;
```
bracket initializers everywhere

michaelpantic commented on time_stamper_cv/include/Common.h:
```diff
@@ -0,0 +1,16 @@
+  double angle;
+};
+
+//Forward declarations
```
extend comment as to why we need them

michaelpantic commented on time_stamper_cv/src/Calibration.cpp:
```diff
@@ -0,0 +1,142 @@
+#include "ros/ros.h"
+
+Calibration::Calibration(const CalibrationConfig &cfg) {
+  convex_shape_ = new ConvexShape(cfg.tolerance);
```
use shared_ptr for all ptrs

michaelpantic commented on time_stamper_cv/src/Calibration.cpp:
```diff
@@ -0,0 +1,142 @@
+
+  PointVector points = ConvertKeyPoints();
+  convex_shape_->Process(points);
+  SetShapeStatus();
```
increase cohesion and decrease coupling by moving shapestatus to convex_shape

michaelpantic commented on time_stamper_cv/src/Calibration.cpp:
```diff
@@ -0,0 +1,142 @@
+  cv::Mat visualization_mat = input_mat.clone();
+  detector_->detect(input_mat, keypoints_);
+
+  SetKeypointStatus();
```
potentially wrap detector in your own class and keep keypoints and their status there

michaelpantic commented on time_stamper_cv/src/Calibration.cpp:
```diff
@@ -0,0 +1,142 @@
+  int number = -1;
+  if (convex_shape_->isShapeValid()) {
+    led_parser_->ProcessImage(input_mat);
+    cv::Mat inv_homography =
```
document why we get the homography and what we are doing with it

michaelpantic commented on time_stamper_cv/src/Calibration.cpp:
```diff
@@ -0,0 +1,142 @@
+  convex_shape_->Process(points);
+  SetShapeStatus();
+
+  int number = -1;
```
what is number?

michaelpantic replied:
ah! move declaration closer to usage

michaelpantic replied:
maybe processImage should return the number? or a result object

michaelpantic commented on time_stamper_cv/src/Calibration.cpp:
```diff
@@ -0,0 +1,142 @@
+  PointVector points{};
+  points.reserve(keypoints_.size());
+
+  std::transform(keypoints_.begin(),
```
add this as a custom cast converter to PointVector

michaelpantic commented on time_stamper_cv/src/Calibration.cpp:
```diff
@@ -0,0 +1,142 @@
+}
+
+void Calibration::SetShapeStatus() {
+  if (convex_shape_->isShapeValid()) {
```
isShapeValid() != isLastShapeValid to detect change

and then construct string based on isShapeValid()

michaelpantic replied:
(same for keypoints)

michaelpantic commented on time_stamper_cv/src/Calibration.cpp:
```diff
@@ -0,0 +1,142 @@
+  cv::putText(visualization_mat, counter_text, cv::Point(s.width * 0.05, s.height * 0.9),
+              cv::FONT_HERSHEY_COMPLEX_SMALL, 0.8, cv::Scalar(255, 0, 0));
+  cv::imshow(OPENCV_WINDOW + std::string(" Visualization"), visualization_mat);
+  cv::waitKey(3);
```
is needed?

maybe cv::waitKey(0) or 1

michaelpantic commented on time_stamper_cv/include/ConvexShape.h:
```diff
@@ -0,0 +1,97 @@
+/**
+ * At the moment only hardcoded shape (@var virtualCorners) is supported.
+ * ConvexShape should support all polygons in future.
+ */
```
add explanation how this class works (the sorting etc)

michaelpantic commented on time_stamper_cv/src/ConvexShape.cpp:
```diff
@@ -0,0 +1,135 @@
+  return Filter(expected_value - tolerance_, expected_value + tolerance_, actual_value);
+}
+
+bool ConvexShape::Filter(double min, double max, double value) {
```
better: inline function not assigned with class or in some common class

michaelpantic commented on time_stamper_cv/src/ConvexShape.cpp:
```diff
@@ -0,0 +1,135 @@
+  }
+}
+
+bool ConvexShape::ToleranceFilter(int expected_value, double actual_value) const {
```
int vs double

michaelpantic replied:
naming: Filter in robotics is very often used, so better something about "InRange" or "Bounded"

michaelpantic commented on time_stamper_cv/src/ConvexShape.cpp:
```diff
@@ -0,0 +1,135 @@
+  return hull_.size() >= 4;
+}
+
+Point2fVector ConvexShape::getVirtualCorners(int multiplier) {
```
use linear shift instead of multiplication

michaelpantic replied:
rename to scaling
Maybe use scaling matrix?

michaelpantic commented on time_stamper_cv/src/ConvexShape.cpp:
```diff
@@ -0,0 +1,135 @@
+  for (auto &point_c : hull_) {
+    double angle = Trigonometry::CalcInnerAngle(point_a, point_c, point_b);
+
+    if (angle < (180 - tolerance_)) {
```
add comment

michaelpantic commented on time_stamper_cv/src/ConvexShape.cpp:
```diff
@@ -0,0 +1,135 @@
+    }
+  }
+
+  if (count != 2) {
```
add comment 
(something like "pattern not found")

michaelpantic commented on time_stamper_cv/src/ConvexShape.cpp:
```diff
@@ -0,0 +1,135 @@
+}
+
+bool ConvexShape::calculateSortedPointAngles() {
+  point_angles_sorted_ = calculatePointAngles();
```
could be solved using autocorrelation

michaelpantic commented on time_stamper_cv/src/LedParser.cpp:
```diff
@@ -0,0 +1,93 @@
+LedParser::LedParser(LedRowConfig  led_row_config, int image_crop_size)
+: led_row_config_(std::move(led_row_config)), size_(image_crop_size) {
+
+  cv::Mat kernel = cv::Mat(image_crop_size, image_crop_size, CV_8UC1);
+
+  int half_size = image_crop_size / 2;
+  cv::circle(kernel, cv::Point(half_size, half_size), half_size,
+             cv::Scalar(255, 255, 255), -1);
+
+ kernel_normalized_ = kernel / 255;
```
comment on idea and kernel

michaelpantic commented on time_stamper_cv/src/LedParser.cpp:
```diff
@@ -0,0 +1,93 @@
+  cv::Point2f led_pos = cfg.first_led_pos;
+
+  for (int i = 1; i <= cfg.amount; i++) {
+    led_row.emplace_back((led_pos.x + (cfg.next_led.val[0] * (i * 1.0)) * cfg.multiplier),
```
use casts

michaelpantic commented on time_stamper_cv/src/LedParser.cpp:
```diff
@@ -0,0 +1,93 @@
+  cv::Point3_<float> led_transformed = led_row_.at(index);
+
+  cv::Point2f led_pos = Normalize(led_transformed);
+  int row_begin = (int) (led_pos.x - (size_ / 2.0));
```
use cv::point

michaelpantic commented on time_stamper_cv/src/LedParser.cpp:
```diff
@@ -0,0 +1,93 @@
+  int row_begin = (int) (led_pos.x - (size_ / 2.0));
+  int col_begin = (int) (led_pos.y - (size_ / 2.0));
+
+  if ((col_begin + size_) > image_.rows || row_begin + size_ > image_.cols) {
```
use cv::point

michaelpantic replied:
dont forget to check against 0/0

michaelpantic commented on time_stamper_cv/src/LedParser.cpp:
```diff
@@ -0,0 +1,93 @@
+    return -1;
+  }
+
+  if (row_begin > 0 && col_begin > 0) {// &&row_begin + size_ < image_.rows && col_begin + size_ < image_.cols) {
```
clean comment

michaelpantic commented on time_stamper_cv/src/LedParser.cpp:
```diff
@@ -0,0 +1,93 @@
+    cv::Mat cropped = image_(led_rect);
+
+    //average - cv::sum(...) > 0 is always true
+    double normalization = 1.0 / cv::sum(kernel_normalized_)[0];
```
add normalization directly to kernel

michaelpantic commented on time_stamper_cv/src/LedParser.cpp:
```diff
@@ -0,0 +1,93 @@
+  return -1;
+}
+
+bool LedParser::isLedOn(int index, int min_brightness) {
```
min_brightness could be a float

michaelpantic commented on time_stamper_cv/src/Trigonometry.cpp:
```diff
@@ -0,0 +1,18 @@
+#include "Trigonometry.h"
+
+double Trigonometry::CalcDistance(const cv::Point &a, const cv::Point &b) {
+  return sqrt(std::pow((a.x - b.x), 2) + (std::pow((a.y - b.y), 2)));
```
try using cv::norm(a-b)

michaelpantic commented on time_stamper_cv/src/Trigonometry.cpp:
```diff
@@ -0,0 +1,18 @@
+double Trigonometry::CalcInnerAngle(
+    const cv::Point &a, const cv::Point &b, const cv::Point &target
+) {
+  double distance_c = CalcDistance(a, b);
```
name distance_ab

michaelpantic commented on time_stamper_cv/src/Trigonometry.cpp:
```diff
@@ -0,0 +1,18 @@
+}
+
+double Trigonometry::LawOfCos(double a, double b, double c) {
+  return acos((c * c - (a * a + b * b)) / -(2 * a * b)) * 180 / M_PI;
```
check if a or b are 0 (avoid div/0)

michaelpantic commented on time_stamper_cv/test/test_convex_shape.cc:
```diff
@@ -0,0 +1,130 @@
+  EXPECT_EQ(convex_shape_test.Filter(10, 30, 10), true);
+}
+
+TEST(ConvexShapeFilter, TestLowerLimit) {
```
TestLowerTolerance

michaelpantic reviewed (changes requested):
Looks pretty good!
- use const/ref wherever possible
- change all functionnames to be lower camel case (first letter)
- change file names to google standard
- create UML diagrams with dependencies

4c3y committed:
Improved const correctness

4c3y committed:
Updated launchfile

4c3y committed:
Renamed node to LedDetectionNode + refactoring

4c3y committed:
Added bracket initializers everywhere

4c3y committed:
Removed unnecessary forward declarations

4c3y committed:
Renamed Node.cpp to LedDetectionNode.cpp, Converted raw pointer to smart pointer

4c3y committed:
Refactored Calibration::SetKeypointStatus and Calibration::SetShapeStatus

4c3y committed:
Converted PointVector to class

4c3y committed:
Added Detector.h, Refactored Calibration

4c3y committed:
Moved SetShapeStatus() to ConvexShape and improved implementation

4c3y committed:
Refactored Calibration,
Moved Filter() to Common.h,
Added missing constructor in PointVector for unittests,
Renamed ToleranceFilter to isInRange,
Added ConvexShape::getInvHomography(),
Refactored LedParser and Trigonometry

4c3y committed:
Changed all function names to be lower camel case. Changed file names according to style guide. Renamed calibration to image_processor

4c3y committed:
Renamed CalibrationConfig to ImageProcessorConfig, Added DependencyGraph.puml

michaelpantic commented on time_stamper_cv/include/common.h:
```diff
@@ -0,0 +1,39 @@
+
+class Common {
+ public:
+  static bool filter(double min, double max, double value) {
```
rename to inRange or so

michaelpantic replied:
params should all be const

michaelpantic commented on time_stamper_cv/include/configuration.h:
```diff
@@ -0,0 +1,45 @@
+ */
+
+struct ImageProcessorConfig {
+  int tolerance{};
```
set default value

michaelpantic commented on time_stamper_cv/include/convex_shape.h:
```diff
@@ -0,0 +1,112 @@
+  PointAngleVector calculatePointAngles();
+  bool calculateSortedPointAngles();
+
+  int tolerance_{};
```
set default value

michaelpantic commented on time_stamper_cv/include/detector.h:
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include "detector.h"
+#include "opencv2/opencv.hpp"
+class detector {
```
Class names should be capitalized CamelCase

michaelpantic replied:
(name should probably be KeyPointDetector or so)

michaelpantic commented on time_stamper_cv/include/detector.h:
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include "detector.h"
```
includes itself

michaelpantic commented on time_stamper_cv/include/led_parser.h:
```diff
@@ -0,0 +1,71 @@
+
+  /**
+   * Generate a point3f vector from LedRowConfig, where the z value is 1.
+   * Used for some matrix operations.
```
:-) Can be called homogenous 2d coordinates

michaelpantic commented on time_stamper_cv/src/convex_shape.cpp:
```diff
@@ -0,0 +1,140 @@
+  }
+}
+
+bool ConvexShape::isInRange(double value1, double value2) const {
```
this hsould probably be isWithinTolerance or similar

michaelpantic commented on time_stamper_cv/src/detector.cpp:
```diff
@@ -0,0 +1,27 @@
+  return keypoints_.empty();
+}
+
+void detector::pollKeyPointStatus(const std::function<void(std::string)>& function) {
```
why not just return a string?

4c3y replied:
Because the std::function is intended to only get called if the keypoint state change.

michaelpantic commented on time_stamper_cv/src/led_parser.cpp:
```diff
@@ -0,0 +1,92 @@
+
+#include <utility>
+
+LedParser::LedParser(LedRowConfig led_row_config, int image_crop_size)
```
hmm LedStateDetector would be a more verbose name

4c3y replied:
Renamed to LedStateParser to not confuse it with KeypointDetector

michaelpantic commented on time_stamper_cv/include/common.h:
```diff
@@ -0,0 +1,39 @@
+    std::copy(keypoints.begin(), keypoints.end(), std::back_inserter(*this));
+  }
+
+  explicit PointVector(std::vector<cv::KeyPoint> keypoints) {
```
param const

4c3y replied:
This can't be const because otherwise the unary operator fails to compile

michaelpantic commented on time_stamper_cv/include/convex_shape.h:
```diff
@@ -0,0 +1,112 @@
+   * Creates ConvexShape object
+   * @param tolerance is used for isInRange().
+   */
+  explicit ConvexShape(int tolerance);
```
param const

4c3y replied:
Added const in cpp file to avoid warnings

michaelpantic commented on time_stamper_cv/include/convex_shape.h:
```diff
@@ -0,0 +1,112 @@
+   * @param scaling
+   * @return Scaled virtualCorners
+   */
+  Point2fVector getVirtualCorners(float scaling = 1.0f);
```
param const

4c3y replied:
Added const in cpp file to avoid warnings

michaelpantic commented on time_stamper_cv/include/convex_shape.h:
```diff
@@ -0,0 +1,112 @@
+   * @return true if in tolerance, otherwise false.
+   */
+
+  bool isInRange(double value1, double value2) const;
```
params const

4c3y replied:
same here

michaelpantic commented on time_stamper_cv/include/image_processor.h:
```diff
@@ -0,0 +1,48 @@
+   * If set to true, a visualization window is opened and updated every frame.
+   * @param visualization
+   */
+  void setVisualization(bool visualization);
```
param const

4c3y replied:
same here

michaelpantic commented on time_stamper_cv/include/led_parser.h:
```diff
@@ -0,0 +1,71 @@
+   * @param min_brightness Minimum brightness, value between 1 and 255.
+   * @return true if average pixel brightness > min_brightness, otherwise false.
+   */
+  bool isLedOn(int index, float min_brightness = 40.0) const;
```
params ocnst

4c3y replied:
same here

michaelpantic reviewed (approved):
Looks good to me!

See comments - please mark them as resolved (if resolved) or comment on them before hitting the merge button :-)

Also - make sure it passes the build before merging

4c3y committed:
Improved const correctness and fixed docs in header + Refactoring

michaelpantic committed:
Added functional header for compatiblity with older gcc
